### PR TITLE
Correct field name for stripped state events when knocking. `knock_state_events` -> `knock_room_state`

### LINKED ISSUE #14088: When knocking, Synapse returns and accepts a list of stripped state events in an incorrectly named field
When a homeserver knocks on a room over federation, it calls the [` PUT /_matrix/federation/v1/send_knock/{roomId}/{eventId}`](https://spec.matrix.org/v1.4/server-server-api/#put_matrixfederationv1send_knockroomideventid) endpoint on the remote homeserver. The remote homeserver can then return a set of [stripped state events](https://spec.matrix.org/v1.4/client-server-api/#stripped-state) that describe the room, which are passed subsequently to the client.

The response field for a knock's stripped state events is defined in the spec as `knock_room_state`. It looks like it was implemented as `knock_state_events` in https://github.com/matrix-org/synapse/pull/6739/files#diff-996dafa2cabdf043e51a444e5d3c5ba8389d48e6cdab0d469677974bd470d58fR686. A subtle but important difference.

As far as I can tell, this is a problem that's exclusive to Synapse. [Conduit](https://gitlab.com/famedly/conduit) does not implement knocking, [Dendrite](https://github.com/matrix-org/dendrite/) supports knock membership but not initiating a knock.

Assuming we want to support older Synapse's knocking through a newer Synapse and vice versa, a backwards-compatible fix would involve:

* Supporting both field names when extracting stripped state events from a response to [` PUT /_matrix/federation/v1/send_knock/{roomId}/{eventId}`](https://spec.matrix.org/v1.4/server-server-api/#put_matrixfederationv1send_knockroomideventid)
* Including the room's stripped state events under both field names when responding to the above endpoint.

If we didn't implement backwards-compatibility, a user on an old Synapse (or knocking on a new Synapse through an old Synapse) would (assuming knocking was implemented in any clients) see only a room ID in their "pending knocks" list, instead of a room name/avatar/any other metadata.

Alternatively we could opt for changing the field name in the spec and leave Synapse as is, but I'd be adverse as `knock_room_state` currently lines up nicely with `invite_room_state` for [`PUT /_matrix/federation/v2/invite/{roomId}/{eventId} `](https://spec.matrix.org/v1.4/server-server-api/#put_matrixfederationv2inviteroomideventid). It'd also require an MSC.

### PR DESCRIPTION
**Pull request by anoadragon453**
Closes #14088.

This PR:

* Attempts to pull out stripped state events from first a `knock_room_state` field (new name), then a `knock_state_events` field (old name) if present.
* Returns stripped state events under both names if order to support older Synapse instances.

### CODE REVIEW

File: synapse/handlers/federation.py [hunk 1]
```diff
@@ -781,15 +781,28 @@ async def do_knock(
...
+            # See https://github.com/matrix-org/synapse/issues/14088.
+            # TODO: Remove when enough of the ecosystem have upgraded to Synapse v1.xx+.
+            or knock_response.get("knock_state_events")
+        )
+
+        if stripped_room_state is None:
+            raise KeyError(
+                "Missing 'knock_room_state' (or legacy 'knock_state_events') field in "
+                "send_knock response"
+            )
```
Comment (anoadragon453) [inline synapse/handlers/federation.py:803]: I think a different exception type would be more appropriate (and user-friendly) here. But I'm inclined not to change things too much in this PR.

The earlier code would have also raised a `KeyError` in the case the field was not present.

**Review by squahtx — APPROVED** [review]
LGTM! Can we file an issue to remove `knock_state_events`?

**Comment by anoadragon453** [comment]
> Can we file an issue to remove `knock_state_events`?

Good shout! https://github.com/matrix-org/synapse/issues/14151

**Comment by anoadragon453** [comment]
I believe the CI failure is a flake as it seems entirely unrelated to the code changes.

**Pull request merged**